### ISSUE TITLE
Incompatibility SF4 Service Container?

### ISSUE THREAD

**Issue by shvlv** [sonata-project/SonataUserBundle#1023]
### Environment

laravel/homestead (virtualbox, 5.2.0)

#### Sonata packages

```
sonata-project/admin-bundle              3.35.0 3.35.1 The missing Symfony Admin Generator
sonata-project/block-bundle              3.12.1 3.12.1 Symfony SonataBlockBundle
sonata-project/cache                     2.0.1  2.0.1  Cache library
sonata-project/core-bundle               3.9.1  3.9.1  Symfony SonataCoreBundle
sonata-project/datagrid-bundle           2.3.1  2.3.1  Symfony SonataDatagridBundle
sonata-project/doctrine-extensions       1.0.2  1.0.2  Doctrine2 behavioral extensions
sonata-project/doctrine-orm-admin-bundle 3.5.1  3.6.0  Symfony Sonata / Integrate Doctrine ORM into the SonataAdminBundle
sonata-project/easy-extends-bundle       2.5.0  2.5.0  Symfony SonataEasyExtendsBundle
sonata-project/exporter                  1.8.0  1.8.0  Lightweight Exporter library
sonata-project/user-bundle               4.1.1  4.1.1  Symfony SonataUserBundle

```

#### Symfony packages

```
symfony/asset                v4.0.8  v4.0.8  Symfony Asset Component
symfony/cache                v4.0.8  v4.0.8  Symfony Cache component with PSR-6, PSR-16, and tags
symfony/class-loader         v3.4.8  v3.4.8  Symfony ClassLoader Component
symfony/config               v4.0.8  v4.0.8  Symfony Config Component
symfony/console              v4.0.8  v4.0.8  Symfony Console Component
symfony/debug                v4.0.8  v4.0.8  Symfony Debug Component
symfony/dependency-injection v4.0.8  v4.0.8  Symfony DependencyInjection Component
symfony/doctrine-bridge      v4.0.8  v4.0.8  Symfony Doctrine Bridge
symfony/dotenv               v4.0.8  v4.0.8  Registers environment variables from a .env file
symfony/event-dispatcher     v4.0.8  v4.0.8  Symfony EventDispatcher Component
symfony/expression-language  v4.0.8  v4.0.8  Symfony ExpressionLanguage Component
symfony/filesystem           v4.0.8  v4.0.8  Symfony Filesystem Component
symfony/finder               v4.0.8  v4.0.8  Symfony Finder Component
symfony/flex                 v1.0.78 v1.0.78
symfony/form                 v4.0.8  v4.0.8  Symfony Form Component
symfony/framework-bundle     v4.0.8  v4.0.8  Symfony FrameworkBundle
symfony/http-foundation      v4.0.8  v4.0.8  Symfony HttpFoundation Component
symfony/http-kernel          v4.0.8  v4.0.8  Symfony HttpKernel Component
symfony/inflector            v4.0.8  v4.0.8  Symfony Inflector Component
symfony/intl                 v4.0.8  v4.0.8  A PHP replacement layer for the C intl extension that includes additional data from the ICU library.
symfony/maker-bundle         v1.4.1  v1.4.2  Symfony Maker helps you create empty commands, controllers, form classes, tests and more so you can forget about writing boilerplate code.
symfony/monolog-bridge       v4.0.8  v4.0.8  Symfony Monolog Bridge
symfony/monolog-bundle       v3.2.0  v3.2.0  Symfony MonologBundle
symfony/options-resolver     v4.0.8  v4.0.8  Symfony OptionsResolver Component
symfony/orm-pack             v1.0.5  v1.0.5  A pack for the Doctrine ORM
symfony/phpunit-bridge       v3.4.8  v4.0.8  Symfony PHPUnit Bridge
symfony/polyfill-intl-icu    v1.7.0  v1.7.0  Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-mbstring    v1.7.0  v1.7.0  Symfony polyfill for the Mbstring extension
symfony/polyfill-php72       v1.7.0  v1.7.0  Symfony polyfill backporting some PHP 7.2+ features to lower PHP versions
symfony/profiler-pack        v1.0.3  v1.0.3  A pack for the Symfony web profiler
symfony/property-access      v4.0.8  v4.0.8  Symfony PropertyAccess Component
symfony/property-info        v4.0.8  v4.0.8  Symfony Property Info Component
symfony/routing              v4.0.8  v4.0.8  Symfony Routing Component
symfony/security             v4.0.8  v4.0.8  Symfony Security Component
symfony/security-acl         v3.0.1  v3.0.1  Symfony Security Component - ACL (Access Control List)
symfony/security-bundle      v4.0.8  v4.0.8  Symfony SecurityBundle
symfony/serializer           v4.0.8  v4.0.8  Symfony Serializer Component
symfony/serializer-pack      v1.0.1  v1.0.1  A pack for the Symfony serializer
symfony/stopwatch            v4.0.8  v4.0.8  Symfony Stopwatch Component
symfony/swiftmailer-bundle   v3.2.2  v3.2.2  Symfony SwiftmailerBundle
symfony/templating           v4.0.8  v4.0.8  Symfony Templating Component
symfony/translation          v4.0.8  v4.0.8  Symfony Translation Component
symfony/twig-bridge          v4.0.8  v4.0.8  Symfony Twig Bridge
symfony/twig-bundle          v4.0.8  v4.0.8  Symfony TwigBundle
symfony/validator            v4.0.8  v4.0.8  Symfony Validator Component
symfony/var-dumper           v4.0.8  v4.0.8  Symfony mechanism for exploring and dumping PHP variables
symfony/web-profiler-bundle  v4.0.8  v4.0.8  Symfony WebProfilerBundle
symfony/yaml                 v4.0.8  v4.0.8  Symfony Yaml Component

```

#### PHP version

```
PHP Version 7.1.7-1+ubuntu16.04.1+deb.sury.org+1
```

## Subject
Probably error when using Service Container in SF4.

## Steps to reproduce
Try Forgotten password functionality and go to /admin/resetting/send-email.

## Expected results
Sending email.

## Actual results
Exeption:
```
Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException:
The "fos_user.util.token_generator" service or alias has been removed or inlined when the container was compiled. You should either make it public, or stop using the container directly and use dependency injection instead.

  at vendor/symfony/dependency-injection/Container.php:259
  at Symfony\Component\DependencyInjection\Container->get('fos_user.util.token_generator')
     (vendor/symfony/framework-bundle/Controller/ControllerTrait.php:61)
  at Symfony\Bundle\FrameworkBundle\Controller\Controller->get('fos_user.util.token_generator')
     (vendor/sonata-project/user-bundle/src/Controller/AdminResettingController.php:66)
  at Sonata\UserBundle\Controller\AdminResettingController->sendEmailAction(object(Request))
     (vendor/symfony/http-kernel/HttpKernel.php:149)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/symfony/http-kernel/HttpKernel.php:66)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/symfony/http-kernel/Kernel.php:190)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (public/index.php:37)

```
When look at AdminResettingController, I found next code:

```
$tokenGenerator = $this->get('fos_user.util.token_generator');
```
but services in SF4 is private by default so that code does't work.
I add 
```
fos_user.util.token_generator:
     class: 'FOS\UserBundle\Util\TokenGenerator'
     public: true
```
It work, but 2 problem. Not very good reassign all FOS services again. Even do this, need also add arguments.

It is bug or need additional config? Is workaround exist?

**Comment by kunicmarko20** [comment]
I would say that we make it public in sonata config and we think about refactorng bundles to stop getting from container, cc @sonata-project/contributors

**Comment by greg0ire** [comment]
>I would say that we make it public in sonata config

how can we achieve this?

**Comment by kunicmarko20** [comment]
Can't we just register service and add public true, like in example from issue?

**Comment by shvlv** [comment]
Like temporary fix I find this decision

services.yaml
```
services:
    # Fix Sonata User Bundle controller private service getter
    fos_user.util.token_generator:
        alias: 'fos_user.util.token_generator.default'
        public: true
```
Kernel.php
```
class Kernel extends BaseKernel implements CompilerPassInterface {
	use MicroKernelTrait;
...
public function process(ContainerBuilder $container) {
		// Fix Sonata User Bundle controller private service getter
		foreach ($this->getSonataNeedPrivateServices() as $service) {
			$container->getDefinition($service)->setPublic(true);
		}
	}

protected function getSonataNeedPrivateServices() {
      return [
		'fos_user.resetting.form.factory',
		'fos_user.security.login_manager'
	];
      }
}
```

**Comment by 4ssil** [comment]
@shvlv using the proposed solution above to make the services public from the Kernel didn't work for me, I guess the container couldn't load the services because they're private ?

I had to redefine the services in my services.yaml to get the whole password reset process working.

**services.yaml**
```
services:
    fos_user.util.token_generator:
        alias: 'fos_user.util.token_generator.default'
        public: true
    fos_user.resetting.form.factory:
        class: 'FOS\UserBundle\Form\Factory\FormFactory'
        arguments: ['@form.factory','%fos_user.resetting.form.name%', '%fos_user.resetting.form.type%', '%fos_user.resetting.form.validation_groups%']
        public: true
    fos_user.security.login_manager:
        class: FOS\UserBundle\Security\LoginManager
        arguments: [@security.token_storage','@security.user_checker','@security.authentication.session_strategy','@request_stack',~]
        public: true
```

**Comment by greg0ire** [comment]
If you want to help, you can try using dependency injection inside the controllers. Here is something very similar I did for a command: https://github.com/sonata-project/SonataUserBundle/pull/1074

**Comment by phansys** [comment]
Thank you for reporting this issue.
I'm closing this since #1053 was already merged and its changes are released in version [4.3.0](https://github.com/sonata-project/SonataUserBundle/releases/tag/4.3.0).
Please, feel free to reopen if you think there is something missing there.
Regards!
